snap-csv-dataio/src/main/java/org/esa/snap/csv/dataio/writer/CsvProductWriter.java

```diff
@@ -132,61 +136,41 @@ public void writeBandRasterData(Band sourceBand, int sourceOffsetX, int sourceOf
         if (productWritten) {
             return;
         }
-        final Band[] bands = getSourceProduct().getBands();
-        final DataBuffer[] bandDataBuffers = new DataBuffer[bands.length];
-        final int[] bandTypes = new int[bands.length];
-        for (int i = 0; i < bands.length; i++) {
-            final Band band = bands[i];
-            bandDataBuffers[i] = band.getGeophysicalImage().getData().getDataBuffer();
-            bandTypes[i] = bandDataBuffers[i].getDataType();
-        }
-
+        final RasterDataNode[] bands = getSourceProduct().getBands();
         final TiePointGrid[] tiePointGrids = getSourceProduct().getTiePointGrids();
-        final DataBuffer[] tpgDataBuffers = new DataBuffer[tiePointGrids.length];
-        final int[] tpgTypes = new int[tiePointGrids.length];
-        for (int i = 0; i < tiePointGrids.length; i++) {
-            final TiePointGrid tpg = tiePointGrids[i];
-            tpgDataBuffers[i] = tpg.getGeophysicalImage().getData().getDataBuffer();
-            tpgTypes[i] = tpgDataBuffers[i].getDataType();
+        final RasterDataNode[] rasterDataNodes = Arrays.copyOf(bands, bands.length + tiePointGrids.length);
+        System.arraycopy(tiePointGrids, 0, rasterDataNodes, bands.length, tiePointGrids.length );
+
+        final int[] dataTypes = new int[bands.length];
+
+        final Raster[] dataRasters = new Raster[rasterDataNodes.length];
+        // todo: For big products this will not work because the hole scene is allocated. But no one wants to
```

Comment:
Rather than having a todo here, we should throw an exception for this case in writeProductNodesImpl(). If you keep the todo, change 'hole' to 'whole' and 'adressed' to 'addressed'.

Reply:
I keep my typos and also the todo.
At which point do you want to start to throw an exception? It depends on the amount of available memory if it works or not.

Reply:
We could ask for that ( Runtime.getRuntime().freeMemory() ), estimate how much we need and compare.

Reply:
Yes, possible, but it would be better if the data would be written tile wise.
I would like to save additional work for an interim solution.

Reply:
typos fixed

Reply:
Well, then we should also keep in mind to make the writer able to write multi-size products ... and the reader to read multi-size products.